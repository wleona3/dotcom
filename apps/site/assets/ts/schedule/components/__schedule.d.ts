import { TypedRoutes } from "../../stop/components/__stop";
import {
  Alert,
  Route,
  PredictedOrScheduledTime,
  EnhancedRoute,
  DirectionId,
  Service,
  Schedule,
  RoutePattern,
  Shape,
  Stop,
  ClosedStopInfo
} from "../../__v3api";

export interface EnhancedRoutePattern extends RoutePattern {
  shape_id: string;
  shape_priority: number;
  headsign: string;
}

export interface ShapesById {
  [key: string]: Shape;
}

export interface RoutePatternsByDirection {
  [key: string]: EnhancedRoutePattern[];
}

export interface ServiceInSelector extends Service {
  "default_service?": boolean;
}

export interface SchedulePageData {
  connections: TypedRoutes[];
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
  fares: Fare[];
  fare_link: string;
  holidays: Holiday[];
  route: EnhancedRoute;
  schedule_note: ScheduleNote | null;
  services: ServiceInSelector[];
  stops: SimpleStopMap;
  direction_id: DirectionId;
  route_patterns: RoutePatternsByDirection;
  line_diagram: LineDiagramStop[];
  today: string;
  variant: string | null;
}

interface StopData {
  branch: string | null;
  type: "line" | "merge" | "stop" | "terminus" | null;
  "has_disruption?": boolean;
}

export interface LineDiagramStop {
  stop_data: StopData[];
  route_stop: RouteStop;
  alerts: Alert[];
}

export type CrowdingType = "not_crowded" | "some_crowding" | "crowded" | null;

export interface LineDiagramVehicle {
  id: string;
  status: "in_transit" | "incoming" | "stopped";
  crowding: CrowdingType;
  tooltip: string;
}

interface RouteStopRoute extends Route {
  "custom_route?": boolean;
}
export interface RouteStop {
  id: string;
  name: string;
  zone: string | null;
  branch: string | null;
  station_info: Stop & { parent_id: string | null; child_ids: string[] };
  route: RouteStopRoute | null;
  connections: RouteStopRoute[];
  stop_features: string[];
  "is_terminus?": boolean;
  "is_beginning?": boolean;
  closed_stop_info: ClosedStopInfo | null;
}

export interface SimpleStopMap {
  [key: string]: SimpleStop[];
}

export interface ServiceSchedule {
  [key: string]: ServiceScheduleInfo;
}

export interface ScheduleInfo {
  schedules: ScheduleWithFare[];
  duration: string;
  route_pattern_id: string;
}

export interface ScheduleWithFare extends Schedule {
  price: string;
  fare_link: string;
}

export interface ServiceScheduleByTrip {
  [key: string]: ScheduleInfo;
}

export interface ServiceScheduleInfo {
  by_trip: ServiceScheduleByTrip;
  trip_order: string[];
}

export interface ScheduleNote {
  peak_service: string;
  offpeak_service: string;
  exceptions: ServiceException[];
  alternate_text: string | null;
}

export interface ServiceException {
  type: string;
  service: string;
}

export interface SimpleStop {
  id: string;
  name: string;
  is_closed: boolean;
  zone: string | null;
}

export interface SchedulePDF {
  title: string;
  url: string;
}

export interface Fare {
  title: string;
  price: string;
}

export interface Holiday {
  name: string;
  date: string;
}

export interface StopPrediction {
  headsign: string;
  route: Route;
  prediction: PredictedOrScheduledTime;
  train_number: string;
}

export type SelectedOrigin = string | null;
export interface UserInput {
  route: string;
  origin: string;
  date: string;
  direction: DirectionId;
}
