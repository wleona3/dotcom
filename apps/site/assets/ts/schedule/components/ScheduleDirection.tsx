import React, { ReactElement, useReducer, useEffect, Dispatch } from "react";
import { DirectionId, EnhancedRoute } from "../../__v3api";
import {
  RoutePatternsByDirection,
  LineDiagramStop,
  SimpleStopMap,
  ServiceInSelector,
  ScheduleNote as ScheduleNoteType,
  EnhancedRoutePattern
} from "./__schedule";
import ScheduleDirectionMenu from "./direction/ScheduleDirectionMenu";
import ScheduleDirectionButton from "./direction/ScheduleDirectionButton";
import { reducer as fetchReducer } from "../../helpers/fetch";
import { menuReducer, FetchAction } from "./direction/reducer";
import { MapData, StaticMapData } from "../../leaflet/components/__mapdata";
import Map from "./Map";
import LineDiagramAndStopListPage from "./line-diagram/LineDiagram";
import { isABusRoute, isACommuterRailRoute } from "../../models/route";

export interface Props {
  route: EnhancedRoute;
  directionId: DirectionId;
  routePatternsByDirection: RoutePatternsByDirection;
  mapData?: MapData;
  staticMapData?: StaticMapData;
  lineDiagram: LineDiagramStop[];
  services: ServiceInSelector[];
  stops: SimpleStopMap;
  today: string;
  scheduleNote: ScheduleNoteType | null;
  busVariantId: string | null;
}

export const fetchMapData = (
  routeId: string,
  directionId: DirectionId,
  currentRoutePatternIdForData: string | undefined,
  dispatch: Dispatch<FetchAction>
): Promise<void> => {
  dispatch({ type: "FETCH_STARTED" });
  const baseURL = `/schedules/map_api?id=${routeId}&direction_id=${directionId}`;
  const url = currentRoutePatternIdForData
    ? `${baseURL}&shape_id=${currentRoutePatternIdForData}`
    : baseURL;

  return (
    window.fetch &&
    window
      .fetch(url)
      .then(response => {
        if (response.ok) return response.json();
        throw new Error(response.statusText);
      })
      .then(json => dispatch({ type: "FETCH_COMPLETE", payload: json }))
      // @ts-ignore
      .catch(() => dispatch({ type: "FETCH_ERROR" }))
  );
};

export const fetchLineData = (
  routeId: string,
  directionId: DirectionId,
  currentRoutePatternIdForData: string | undefined,
  dispatch: Dispatch<FetchAction>
): Promise<void> => {
  dispatch({ type: "FETCH_STARTED" });

  const baseURL = `/schedules/line_api?id=${routeId}&direction_id=${directionId}`;
  const url = currentRoutePatternIdForData
    ? `${baseURL}&route_pattern=${currentRoutePatternIdForData}`
    : baseURL;

  return (
    window.fetch &&
    window
      .fetch(url)
      .then(response => {
        if (response.ok) return response.json();
        throw new Error(response.statusText);
      })
      .then(json => dispatch({ type: "FETCH_COMPLETE", payload: json }))
      // @ts-ignore
      .catch(() => dispatch({ type: "FETCH_ERROR" }))
  );
};

const ScheduleDirection = ({
  route,
  directionId,
  routePatternsByDirection,
  mapData,
  staticMapData,
  lineDiagram,
  busVariantId
}: Props): ReactElement<HTMLElement> => {
  const routePatternsInCurrentDirection = routePatternsByDirection[directionId];
  const defaultRoutePattern =
    routePatternsInCurrentDirection.find(
      routePattern => routePattern.id === busVariantId
    ) || routePatternsInCurrentDirection.slice(0, 1)[0];

  const reverseDirection = directionId === 0 ? 1 : 0;
  const directionIsChangeable = route.direction_names[reverseDirection] != null;

  const [state, dispatch] = useReducer(menuReducer, {
    routePattern: defaultRoutePattern,
    directionId,
    routePatternsByDirection,
    routePatternMenuOpen: false,
    routePatternMenuAll: false,
    itemFocus: null
  });

  // const [mapState, dispatchMapData] = useReducer(fetchReducer, {
  //   data: mapData,
  //   isLoading: false,
  //   error: false
  // });

  // To distinguish CR shuttles and branches and multi-route trips, we have to filter in a particular order:
  //    1. Filter by route_id - we only want to show the primary path, not the multi-route trips
  //    2. Filter by typicality
  //    3. If there are multiple route_patterns with different shape values, then filter out shuttles (which usually have priority = -1)
  let currentShapes;
  let currentStops;
  if (isABusRoute(route)) {
    currentShapes = [state.routePattern.shape_id];
    currentStops = state.routePattern.stop_ids;
  } else if (isACommuterRailRoute(route)) {
    const currentPatterns = routePatternsInCurrentDirection
      .filter(pattern => pattern.route_id === route.id)
      .reduce((result, current) => {
        if (result.length === 0 || current.typicality < result[0].typicality)
          return [current];
        if (current.typicality === result[0].typicality) {
          if (result[0].shape_priority < 0 && current.shape_priority > 0)
            return [current];
          if (current.shape_priority < 0 && result[0].shape_priority > 0)
            return result;
          return result.concat(current);
        }
        return result;
      }, [] as EnhancedRoutePattern[]);

    currentShapes = currentPatterns.map(pattern => pattern.shape_id);
    currentStops = currentPatterns.reduce(
      (acc, cur) => acc.concat(cur.stop_ids),
      [] as string[]
    );
  } else {
    currentShapes = routePatternsInCurrentDirection.map(
      pattern => pattern.shape_id
    );
    currentStops = routePatternsInCurrentDirection.reduce(
      (acc, cur) => acc.concat(cur.stop_ids),
      [] as string[]
    );
  }

  const currentRoutePatternIdForData =
    isABusRoute(route) && routePatternsInCurrentDirection.length > 1
      ? state.routePattern.id
      : undefined;

  // useEffect(
  //   () => {
  //     if (!staticMapData && route.id !== "Orange") {
  //       fetchMapData(
  //         route.id,
  //         state.directionId,
  //         currentRoutePatternIdForData,
  //         dispatchMapData
  //       );
  //     }
  //   },
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  //   [route, state.directionId, busVariantId, staticMapData]
  // );

  const [lineState, dispatchLineData] = useReducer(fetchReducer, {
    data: lineDiagram,
    isLoading: false,
    error: false
  });

  useEffect(() => {
    fetchLineData(
      route.id,
      state.directionId,
      currentRoutePatternIdForData,
      dispatchLineData
    );
  }, [route, state.directionId, busVariantId, currentRoutePatternIdForData]);

  return (
    <>
      <div className="m-schedule-direction">
        <div id="direction-name" className="m-schedule-direction__direction">
          {route.direction_names[state.directionId]}
        </div>
        <ScheduleDirectionMenu
          route={route}
          directionId={state.directionId}
          routePatternsByDirection={routePatternsByDirection}
          selectedRoutePatternId={state.routePattern.id}
          menuOpen={state.routePatternMenuOpen}
          showAllRoutePatterns={state.routePatternMenuAll}
          itemFocus={state.itemFocus}
          dispatch={dispatch}
        />
        {directionIsChangeable ? (
          <ScheduleDirectionButton dispatch={dispatch} />
        ) : null}
      </div>
      {/* {!staticMapData && mapData && (
        <Map
          channel={`vehicles:${route.id}:${state.directionId}`}
          data={mapData}
          currentShapes={currentShapes}
          currentStops={currentStops}
        />
      )} */}
      {staticMapData && (
        <>
          <img
            src={staticMapData.img_src}
            alt={`${route.name} route map`}
            className="img-fluid"
          />
          <a
            href={staticMapData.pdf_url}
            target="_blank"
            rel="noopener noreferrer"
          >
            <i className="fa fa-search-plus" aria-hidden="true" />
            View map as a PDF
          </a>
        </>
      )}
      {lineState.data && lineState.data[0] && (
        <LineDiagramAndStopListPage
          lineDiagram={lineState.data}
          route={route}
          directionId={state.directionId}
        />
      )}
    </>
  );
};

export default ScheduleDirection;
