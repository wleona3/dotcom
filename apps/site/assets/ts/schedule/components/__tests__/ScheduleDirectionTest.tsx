import React, { PropsWithChildren } from "react";
import { Provider } from "react-redux";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import {
  closeRoutePatternMenuAction,
  menuReducer as reducer,
  showAllRoutePatternsAction,
  State
} from "../direction/reducer";
import ScheduleDirection, {
  fetchMapData,
  fetchLineData
} from "../ScheduleDirection";
import { EnhancedRoute } from "../../../__v3api";
import { MapData, StaticMapData } from "../../../leaflet/components/__mapdata";
import {
  ShapesById,
  LineDiagramStop,
  RoutePatternsByDirection
} from "../__schedule";
import lineDiagramData from "./test-data/lineDiagramData.json"; // Not a full line diagram
import * as routePatternsByDirectionData from "./test-data/routePatternsByDirectionData.json";
import { createScheduleStore } from "../../store/schedule-store";

const body =
  '<div id="body-wrapper"><div id="react-root"></div><div id="map-root"></div></div>';

const lineDiagram = lineDiagramData as LineDiagramStop[];

const stops = lineDiagram.map(({ route_stop }) => ({
  name: route_stop.name,
  id: route_stop.id,
  is_closed: false,
  zone: route_stop.zone || null
}));

const route = {
  type: 3,
  name: "route 1",
  long_name: "route 1 long name",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
} as EnhancedRoute;

const oneDirectionRoute = {
  type: 3,
  name: "route 2",
  long_name: "route 2 long name",
  id: "route-2",
  direction_names: {
    0: null,
    1: "Inbound"
  },
  direction_destinations: {
    0: null,
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
} as EnhancedRoute;

const directionId = 1;
const routePatternsByDirection = routePatternsByDirectionData as RoutePatternsByDirection;

const shapesById = {
  "shape-1": {
    stop_ids: ["stop"],
    priority: 3,
    polyline: "xyz",
    name: "Shape 1",
    id: "shape-1",
    direction_id: 0
  },
  "shape-2": {
    stop_ids: ["stop"],
    priority: 3,
    polyline: "xyz",
    name: "Shape 2",
    id: "shape-2",
    direction_id: 1
  },
  "shape-3": {
    stop_ids: ["stop"],
    priority: 3,
    polyline: "xyz",
    name: "Shape 3",
    id: "shape-3",
    direction_id: 0
  }
} as ShapesById;

const state = {
  routePattern: routePatternsByDirection["0"][0],
  shape: shapesById["shape-1"],
  directionId: 0,
  shapesById: shapesById,
  routePatternsByDirection: routePatternsByDirection,
  routePatternMenuOpen: false,
  routePatternMenuAll: false,
  itemFocus: null
} as State;

/* eslint-disable camelcase */
const mapData: MapData = {
  zoom: 16,
  width: 600,
  tile_server_url: "https://mbta-map-tiles-dev.s3.amazonaws.com",
  polylines: [],
  markers: [
    {
      icon: "vehicle-bordered-expanded",
      id: "vehicle-R-545CDFC5",
      latitude: 42.39786911010742,
      longitude: -71.13092041015625,
      rotation_angle: 90,
      tooltip_text: "Alewife train is on the way to Alewife",
      tooltip: null
    },
    {
      icon: "stop-circle-bordered-expanded",
      id: "stop-place-alfcl",
      latitude: 42.395428,
      longitude: -71.142483,
      rotation_angle: 0,
      tooltip: null,
      tooltip_text: "Alewife"
    }
  ],
  stop_markers: [
    {
      icon: "stop-circle-bordered-expanded",
      id: "stop-place-alfcl",
      latitude: 42.395428,
      longitude: -71.142483,
      rotation_angle: 0,
      tooltip: null,
      tooltip_text: "Alewife"
    }
  ],
  height: 600,
  default_center: {
    longitude: -71.05891,
    latitude: 42.360718
  }
};

const staticMapData: StaticMapData = {
  img_src: "http://example.com/map.png",
  pdf_url: "http://example.com/map.pdf"
};
/* eslint-enable camelcase */

const getComponent = () => (
  <ScheduleDirection
    route={route}
    directionId={directionId}
    routePatternsByDirection={routePatternsByDirection}
    mapData={mapData}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId={null}
  />
);

const getSingleDirectionComponent = () => (
  <ScheduleDirection
    route={oneDirectionRoute}
    directionId={directionId}
    routePatternsByDirection={routePatternsByDirection}
    mapData={mapData}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId={null}
  />
);

const getSubwayComponent = () => (
  <ScheduleDirection
    mapData={mapData}
    route={{ ...route, type: 1 }}
    directionId={directionId}
    routePatternsByDirection={routePatternsByDirection}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId={null}
  />
);

const getCRComponent = () => (
  <ScheduleDirection
    route={{ ...route, type: 2 }}
    directionId={directionId}
    routePatternsByDirection={routePatternsByDirection}
    mapData={mapData}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId={null}
  />
);

const getStaticMapComponent = () => (
  <ScheduleDirection
    staticMapData={staticMapData}
    route={{ ...route, type: 4 }}
    directionId={directionId}
    routePatternsByDirection={routePatternsByDirection}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId={null}
  />
);

const getGreenLineComponent = () => {
  const greenRoute: EnhancedRoute = {
    type: 0,
    name: "Green Line",
    long_name: "Green Line",
    id: "Green",
    direction_names: { "0": "Westbound", "1": "Eastbound" },
    direction_destinations: {
      "0": "Boston College / Cleveland Circle / Riverside / Heath Street",
      "1": "Park Street / Government Center / North Station / Lechmere"
    },
    description: "rapid_transit",
    header: "",
    alerts: []
  };

  return (
    <ScheduleDirection
      mapData={mapData}
      route={greenRoute}
      directionId={directionId}
      routePatternsByDirection={routePatternsByDirection}
      lineDiagram={lineDiagram}
      services={[]}
      stops={{ stops }}
      today="2019-12-05"
      scheduleNote={null}
      busVariantId={null}
    />
  );
};

const getVariantComponent = () => (
  <ScheduleDirection
    route={route}
    directionId={0}
    routePatternsByDirection={routePatternsByDirection}
    mapData={mapData}
    lineDiagram={lineDiagram}
    services={[]}
    stops={{ stops }}
    today="2019-12-05"
    scheduleNote={null}
    busVariantId="pattern-3"
  />
);
const store = createScheduleStore(0);
// redux store/provider
function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
  return <Provider store={store}>{children}</Provider>;
}
function renderWithProvider(ui: React.ReactElement) {
  return render(ui, { wrapper: Wrapper });
}

afterEach(cleanup);

test("<ScheduleDirection /> renders a bus component", () => {
  const { asFragment } = renderWithProvider(getComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> renders a subway component", () => {
  const { asFragment } = renderWithProvider(getSubwayComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> renders a CR component", () => {
  const { asFragment } = renderWithProvider(getCRComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> can render green line", () => {
  const { asFragment } = renderWithProvider(getGreenLineComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> respects the initially selected pattern ID, if specified", () => {
  const { asFragment } = renderWithProvider(getVariantComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> renders with a static map", () => {
  const { asFragment } = renderWithProvider(getStaticMapComponent());
  expect(asFragment()).toMatchSnapshot();
});

test("<ScheduleDirection /> not allow changing direction when no route patterns", () => {
  const { container, asFragment } = renderWithProvider(
    getSingleDirectionComponent()
  );
  expect(asFragment()).toMatchSnapshot();
  expect(container.querySelector(".m-schedule-direction__button")).toBeNull();
});

test("<ScheduleDirection /> can change direction", () => {
  window.history.replaceState = jest.fn();
  renderWithProvider(getComponent());
  expect(screen.queryByText("Inbound", { exact: false })).toBeTruthy();
  expect(screen.queryByText("Outbound", { exact: false })).toBeFalsy();
  fireEvent.click(screen.getByText("Change Direction", { exact: false }));

  expect(screen.queryByText("Inbound", { exact: false })).toBeFalsy();
  expect(screen.queryByText("Outbound", { exact: false })).toBeTruthy();
  expect(window.history.replaceState).toBeCalledWith(
    {},
    "",
    "/?schedule_direction%5Bdirection_id%5D=0&schedule_direction%5Bvariant%5D=pattern-1"
  );
});

test("<ScheduleDirection /> can change route pattern", () => {
  window.history.replaceState = jest.fn();
  renderWithProvider(getComponent());
  fireEvent.click(screen.getByText("Change Direction", { exact: false })); // shows clickable route pattern options

  fireEvent.click(screen.getByRole("button", { name: "Pattern 1" }));
  expect(window.history.replaceState).toBeCalledWith(
    {},
    "",
    "/?schedule_direction%5Bdirection_id%5D=0&schedule_direction%5Bvariant%5D=pattern-1"
  );

  fireEvent.click(
    screen.getByRole("menuitem", {
      name: "Pattern 3 from Pattern 3, typical route"
    })
  );
  expect(window.history.replaceState).toBeCalledWith(
    {},
    "",
    "/?schedule_direction%5Bdirection_id%5D=0&schedule_direction%5Bvariant%5D=pattern-3"
  );

  // coverage
  fireEvent.click(screen.getByRole("button", { name: "Pattern 3" }));
  fireEvent.keyDown(
    screen.getByRole("menuitem", { name: "Pattern 1 typical route" }),
    { key: "ArrowRight" }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", {
      name: "Pattern 3 from Pattern 3, typical route"
    }),
    { key: "ArrowRight" }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", { name: "click for additional routes" }),
    { key: "ArrowRight" }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", { name: "Pattern 1 typical route" }),
    { key: "ArrowLeft" }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", {
      name: "Pattern 3 from Pattern 3, typical route"
    }),
    { key: "ArrowLeft" }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", {
      name: "Pattern 3 from Pattern 3, typical route"
    }),
    { key: "Tab", shiftKey: true }
  );
  fireEvent.keyDown(
    screen.getByRole("menuitem", {
      name: "Pattern 3 from Pattern 3, typical route"
    }),
    { key: "X" }
  );
  fireEvent.click(
    screen.getByRole("menuitem", { name: "click for additional routes" })
  );
});

// we need to save the original object for later to not affect tests from other files
const realLocation = global.location;

beforeAll(() => {
  const location = {
    ...window.location,
    assign: jest.fn()
  };
  Object.defineProperty(window, "location", {
    writable: true,
    value: location
  });
});

afterAll(() => {
  global.location = realLocation;
});

test("<ScheduleDirection /> can change route for green line", () => {
  const { container } = renderWithProvider(getGreenLineComponent());
  const btn = container.querySelector(
    ".m-schedule-direction__route-pattern--clickable"
  )!;
  fireEvent.click(btn);
  expect(screen.getByRole("menu")).toBeTruthy();
  fireEvent.click(btn);
  expect(screen.queryByRole("menu")).toBeFalsy();
  fireEvent.click(btn);

  fireEvent.click(screen.getByRole("menuitem", { name: "Green Line C" }));
  expect(window.location.assign).toHaveBeenCalledWith(
    "/schedules/Green-C?direction_id=1"
  );
  fireEvent.keyUp(screen.getByRole("menuitem", { name: "Green Line D" }), {
    key: "Enter"
  });
  expect(window.location.assign).toHaveBeenCalledWith(
    "/schedules/Green-D?direction_id=1"
  );
});

it("reducer can change state correctly for closeRoutePatternMenu", () => {
  const previousState = { ...state, routePatternMenuOpen: true } as State;
  const nextState = reducer(previousState, closeRoutePatternMenuAction());
  expect(nextState.routePatternMenuOpen).toEqual(false);
});

it("reducer can change state correctly for showAllRoutePatterns", () => {
  const previousState = { ...state, routePatternMenuAll: false } as State;
  const nextState = reducer(previousState, showAllRoutePatternsAction());
  expect(nextState.routePatternMenuAll).toEqual(true);
});

describe("fetchMapData", () => {
  it("fetches data", () => {
    const spy = jest.fn();
    window.fetch = jest.fn().mockImplementation(
      () =>
        new Promise((resolve: Function) =>
          resolve({
            json: () => mapData,
            ok: true,
            status: 200,
            statusText: "OK"
          })
        )
    );

    return fetchMapData("1", 0, "2", spy).then(() => {
      expect(window.fetch).toHaveBeenCalledWith(
        "/schedules/map_api?id=1&direction_id=0&shape_id=2"
      );
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_STARTED"
      });
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_COMPLETE",
        payload: mapData
      });
    });
  });

  it("fails gracefully if fetch is unsuccessful", () => {
    const spy = jest.fn();
    window.fetch = jest.fn().mockImplementation(
      () =>
        new Promise((resolve: Function) =>
          resolve({
            json: () => "Internal Server Error",
            ok: false,
            status: 500,
            statusText: "INTERNAL SERVER ERROR"
          })
        )
    );

    return fetchMapData("1", 0, "2", spy).then(() => {
      expect(window.fetch).toHaveBeenCalledWith(
        "/schedules/map_api?id=1&direction_id=0&shape_id=2"
      );
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_STARTED"
      });
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_ERROR"
      });
    });
  });
});

describe("fetchLineData", () => {
  it("fetches data", () => {
    const spy = jest.fn();
    window.fetch = jest.fn().mockImplementation(
      () =>
        new Promise((resolve: Function) =>
          resolve({
            json: () => lineDiagramData,
            ok: true,
            status: 200,
            statusText: "OK"
          })
        )
    );

    return fetchLineData("Orange", 1, "1", spy).then(() => {
      expect(window.fetch).toHaveBeenCalledWith(
        "/schedules/line_api?id=Orange&direction_id=1&route_pattern=1"
      );
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_STARTED"
      });
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_COMPLETE",
        payload: lineDiagramData
      });
    });
  });

  it("fails gracefully if fetch is unsuccessful", () => {
    const spy = jest.fn();
    window.fetch = jest.fn().mockImplementation(
      () =>
        new Promise((resolve: Function) =>
          resolve({
            json: () => "Internal Server Error",
            ok: false,
            status: 500,
            statusText: "INTERNAL SERVER ERROR"
          })
        )
    );

    return fetchLineData("Red", 0, "1", spy).then(() => {
      expect(window.fetch).toHaveBeenCalledWith(
        "/schedules/line_api?id=Red&direction_id=0&route_pattern=1"
      );
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_STARTED"
      });
      expect(spy).toHaveBeenCalledWith({
        type: "FETCH_ERROR"
      });
    });
  });
});
