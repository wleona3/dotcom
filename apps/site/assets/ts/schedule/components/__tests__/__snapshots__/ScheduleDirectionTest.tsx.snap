// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduleDirection /> can render green line 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Eastbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern m-schedule-direction__route-pattern--clickable notranslate"
        role="button"
        tabindex="0"
      >
         
        <span
          aria-hidden="true"
          class="notranslate c-svg__icon m-schedule-direction__route-pattern-arrow"
        >
          SVG
        </span>
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> not allow changing direction when no route patterns 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Inbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern notranslate"
      >
        Pattern 2 
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> renders a CR component 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Inbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern notranslate"
      >
        End
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> renders a bus component 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Inbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern notranslate"
      >
        Pattern 2 
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> renders a subway component 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Inbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern notranslate"
      >
        End
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> renders with a static map 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Inbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern notranslate"
      >
        End
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
  <img
    alt="route 1 route map"
    class="img-fluid"
    src="http://example.com/map.png"
  />
  <a
    href="http://example.com/map.pdf"
    rel="noopener noreferrer"
    target="_blank"
  >
    <i
      aria-hidden="true"
      class="fa fa-search-plus"
    />
    View map as a PDF
  </a>
</DocumentFragment>
`;

exports[`<ScheduleDirection /> respects the initially selected pattern ID, if specified 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-direction"
  >
    <div
      class="m-schedule-direction__direction"
      id="direction-name"
    >
      Outbound
    </div>
    <div
      class="js-m-schedule-click-boundary"
    >
      <div
        class="m-schedule-direction__route-pattern m-schedule-direction__route-pattern--clickable notranslate"
        role="button"
        tabindex="0"
      >
        Pattern 3 
        <span
          aria-hidden="true"
          class="notranslate c-svg__icon m-schedule-direction__route-pattern-arrow"
        >
          SVG
        </span>
      </div>
    </div>
    <button
      class="m-schedule-direction__button btn btn-primary"
      type="button"
    >
      <span
        aria-hidden="true"
        class="notranslate m-schedule-direction__icon"
      >
        SVG
      </span>
      Change Direction
    </button>
  </div>
</DocumentFragment>
`;
